mgl32/codegen.go

```diff
@@ -0,0 +1,273 @@
+// Copyright 2014 The go-gl/mathgl Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// +build ignore
+
+// codegen generates go code from templates. Intended to be
+// used with go generate; Also makes mgl64 from mgl32.
+// See the invocation in mgl32/util.go.
+
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"text/template"
+)
+
+type Context struct {
+	Comment      string
+	TemplateName string
+}
+
+type MatrixIter struct {
+	M     int // row
+	N     int // column
+	index int
+}
+
+var mgl64RewriteRules = []string{
+	"mgl32 -> mgl64",
+	"float32 -> float64",
+	"f32 -> f64",
+	"a.Float32 -> a.Float64",
+	"math.MaxFloat32 -> math.MaxFloat64",
+	"math.SmallestNonzeroFloat32 -> math.SmallestNonzeroFloat64",
+}
+
+func main() {
+	flag.Usage = func() {
+		fmt.Println("Usage: codegen -template file.tpl -output file.go")
+		fmt.Println("Usage: codegen -mgl64 [-dir ../mgl64]")
+		flag.PrintDefaults()
+	}
+
+	tplPath := flag.String("template", "file.tpl", "template path")
+	oPath := flag.String("output", "file.go", "output path")
+	mgl64 := flag.Bool("mgl64", false, "make mgl64")
+	mgl64Path := flag.String("dir", "../mgl64", "path to mgl64 location")
+
+	flag.Parse()
+	if flag.NArg() > 0 || flag.NFlag() == 0 {
+		flag.Usage()
+		os.Exit(1)
```

Comment:
Seems good to me, but I'm dubious about moving the stack into the main
package. We were working on having more subpackages rather than fewer.

On Sat, Mar 14, 2015 at 10:06 AM Dmitri Shuralyov <notifications@github.com>
wrote:

> In mgl32/codegen.go
> <https://github.com/go-gl/mathgl/pull/39#discussion_r26439497>:
>
> > +func main() {
> > +	flag.Usage = func() {
> > +		fmt.Println("Usage: codegen -template file.tpl -output file.go")
> > +		fmt.Println("Usage: codegen -mgl64 [-dir ../mgl64]")
> > +		flag.PrintDefaults()
> > +	}
> > +
> > +	tplPath := flag.String("template", "file.tpl", "template path")
> > +	oPath := flag.String("output", "file.go", "output path")
> > +	mgl64 := flag.Bool("mgl64", false, "make mgl64")
> > +	mgl64Path := flag.String("dir", "../mgl64", "path to mgl64 location")
> > +
> > +	flag.Parse()
> > +	if flag.NArg() > 0 || flag.NFlag() == 0 {
> > +		flag.Usage()
> > +		os.Exit(1)
>
> Minor, but this should be os.Exit(2). Exit code 2 is typically used when
> flags are not as expected.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/go-gl/mathgl/pull/39/files#r26439497>.
>